### ISSUE TITLE
Webstorm not Indexing Exports

### ISSUE THREAD

**Issue by xasmx** [kyeotic/raviger#12]
Another hookrouter refugee here.

I'm using Webstorm as my preferred IDE. For some reason it's not able to find the exports from raviger, which causes annoyances such as:
- having to manually import names instead of being able to rely on its auto import feature
- not being able to see the names of the parameters for the functions / suggest what parameters should be passed.

**Comment by kyeotic** [comment]
Welcome! Glad to have another refugee 😁

Unfortunately I don’t know what would cause this. I’ve never used webstorm.
Do you have any insight into what I missed to support this functionality?

**Comment by kyeotic** [comment]
In VS Code the autocomplete is working.

![image](https://user-images.githubusercontent.com/1187432/64914929-ca993e80-d710-11e9-8ba6-7b838eee5141.png)

**Comment by xasmx** [comment]
I spent some time looking further into this. Actually, it looks like Webstorm is indexing the exports ok as long as I'm referring to them from a .js/.jsx file.

I think the problem actually is that my project is in typescript. If I import raviger from a .ts/.tsx file, it won't index the exports. To make I work, I think the project would need to have a .d.ts type/interface declaration file and have in included in package.json.

**Comment by kyeotic** [comment]
Ah, that makes sense. I haven't worked enough with typescript to be familiar with the recommended practice here. Should I submit a declaration to DefinitlyTyped, or bundle it with the npm package?

**Comment by xasmx** [comment]
I've just started using Typescript earlier this year, so I'm probably not the best authority on this :-)
But from my understanding:
- from a typescript users perspective, having them directly in the npm package would be optimal.
- definitelytyped is mostly used when the npm package author hasn't created/added a types declaration directly. Then a third party can add types as a separate package in definitelytyped even if they can't commit to to the npm package directly.

**Comment by kyeotic** [comment]
ok @xasmx Im hoping this typescript declaration file solves your issue. see #13 for details.

install with `npm i raviger@beta` or `npm i raviger@0.5.1-beta.1` and let me know if it works for you

**Comment by xasmx** [comment]
Confirming that it does fix the issue! Thanks.

**Comment by kyeotic** [comment]
@xasmx ok, I will publish it under `0.5.1` shortly

**Comment by kyeotic** [comment]
ok, published under `0.5.1` and removed beta tag

**Comment by xasmx** [comment]
Looks like the types for useRedirect have a typo in the function name:

export function useRedrect(...): void

**Comment by kyeotic** [comment]
@xasmx yep, sorry about that. Fixed under `0.5.2`

**Comment by xasmx** [comment]
Uh oh.. Looks like I didn't do too good of a job testing this. I only tested the indexing of exports in the IDE (which now works), but didn't actually test if my application was still working. Which it isn't.

The current have some issues, such as not dealing with optional parameters and the type for the Link being for props only, not for a component. Anyway, I hacked a bit to try to fix the types and came up with the following:

```
import * as React from 'react'

export function useRoutes(
  routes: Dict<string, (props: Object) => JSX.Element>,
  options?: { basePath?: string; routeProps?: Object; overridePathParams?: boolean }
): JSX.Element

export function useRedirect(
  predicateUrl: string,
  targetUrl: string,
  queryParams?: Object | URLSearchParams,
  replace?: boolean
): void

export interface LinkProps extends React.AnchorHTMLAttributes<HTMLAnchorElement> {
  href: string
}
export const Link : React.FC<LinkProps>;

export function navigate(url: string, replace?: boolean): void
export function navigate(
  url: string,
  query?: Object | URLSearchParams,
  replace?: boolean
): void

export function usePath(basePath?: string): string

export function useBasePath(): string

export function usePopState(
  basePath: string,
  predicate: () => boolean,
  setFn: (path: string) => any
): void

export function useQueryParams<T>(
  parseFn?: (query: string) => T,
  serializeFn?: (query: T) => string
): [T, (query: T) => void]
```

With that, my application managed to get through the typescript compiler again, but now I'm getting some weirdness at runtime. Specifically, when loading the app, I just get:
```
Invalid hook call. Hooks can only be called inside of the body of a function component. This could happen for one of the following reasons:
1. You might have mismatching versions of React and the renderer (such as React DOM)
2. You might be breaking the Rules of Hooks
3. You might have more than one copy of React in the same app
See https://fb.me/react-invalid-hook-call for tips about how to debug and fix this problem.
```

I debugged that for a while, but wasn't able to get it resolved. And unfortunately I don't have time for the next week to really dig into it :-(

Anyway, looks like adding the types might be a bit trickier than what I thought. It might be best to just get rid of them from repository altogether, at least for now. As it stands, the current types will break raviger for any typescript users.

Apparently the recommendation is that types should only be embedded for npms that are written in typescript. And for javascript projects have them exist in the definitelytyped repository. I could probably take a stab at adding them there once I get a bit more time in a week or two.

**Comment by kyeotic** [comment]
That is a common error that comes from using two instances of React, such as when you `npm link` a dependency. Check out [the hooks docs](https://reactjs.org/warnings/invalid-hook-call-warning.html#duplicate-react).

Do you have any references for the typescript type in-the-repo vs in-Definitetly-Typed? I saw some going both ways, but nothing official. I think it makes sense to have them in the repo, they just need to work correctly. I will pull the types and re-release the beta tag with them so you can test them out.

**Comment by kyeotic** [comment]
@xasmx can you try the current beta from #15 with `npm i raviger@beta`

**Comment by xasmx** [comment]
Ah, ok. I was using ```npm link``` to test it, so that's probably why.

Nothing official. I'd guess it's just a common sense recommendation if the project is not expected to be regularly tested against the types for releases. Having them directly in the repo would definitely make things more convenient, if you're willing to maintain them.

Thanks, I'll test it out.

**Comment by kyeotic** [comment]
I’m willing to maintain them, but I do need to figure out how to test them. If you know of any methods to test declaration files I would appreciate it

**Comment by xasmx** [comment]
Seems to work from typescript again.

Using useQueryParams is a bit awkward, as its return type is now specified as Object. To access e.g. 'foo' member from it, I need to first cast it to any:

```
const [queryParams, setQueryParams] : [any, any] = useQueryParams();
const foo = queryParams.foo;
```

While looking at the types, another thing I'm not sure about is the use of 'Dict'. Not familiar with it being a standard interface (though as I said, I'm new to ts so I might be wrong). I've gotten used to declaring map types with ```{ [key : string]: any }```.

**Comment by xasmx** [comment]
Perhaps useQueryParams could use a type such as:
```
export function useQueryParams<T>(
  parseFn?: (query: string) => T,
  serializeFn?: (query: Object) => string
): [T, (query: Object) => void]
```

To allow you to:
```
interface QueryParams {
  foo?: boolean
}

const [queryParams, setQueryParams] = useQueryParams<QueryParams>();
const foo = queryParams.foo;
```

This suggestion is completely untested, for full disclosure :-)

**Comment by kyeotic** [comment]
@xasmx I changed the query type to `{ [key : string]: any }` and added some generic overloads that allow `useQueryParams` to specify the two possible types. Its published under `0.5.4-beta.3` or  the latest `@beta`

**Comment by xasmx** [comment]
Thanks, the latest version seems to be working great!

**Comment by kyeotic** [comment]
@xasmx Do all 3 `useQueryParams` types work for you?

**Comment by xasmx** [comment]
Yep, tested all three as far as the return values go and all seemed good to me. Not really familiar with parseFn and serializeFn, so didn't test them.

**Comment by kyeotic** [comment]
ok, I made a minor refactor. Would you be able to make one last test on `0.5.4-beta.4`? I will publish as `0.5.4` if it works

**Comment by xasmx** [comment]
Retested and all is good.

**Comment by kyeotic** [comment]
Thank you so much for the help on this, I would not have had an easy time testing the types. This is all published under `0.5.4`, and the `beta` tag has been deleted.
